File: Evaluate Reverse Polish Notation/Evaluate_Reverse_Polish_Notation.cpp
// 典型的利用堆栈的题目
// Runtime: 16 ms, faster than 100.00% of C++ online submissions for Evaluate Reverse Polish Notation.
// Memory Usage: 11.5 MB, less than 39.61% of C++ online submissions for Evaluate Reverse Polish Notation.
class Solution 
{
public:
    int evalRPN(vector<string>& tokens) 
    {
        stack<int> memo;
        int res;

        for (auto token : tokens)
        {
            char chr = token[0];

            // 如果是数字的话就压入堆栈
            if (chr >= '0' && chr <= '9' || token.size() > 1)
                memo.push(atoi(token.c_str()));
                
            // 如果是运算符的话就要从堆栈中提取出两个运算数进行运算
            else
            {
                int op1 = memo.top();
                memo.pop();
                int op2 = memo.top();
                memo.pop();

                if (chr == '+')
                    res = op2 + op1;
                else if (chr == '-')
                    res = op2 - op1;
                else if (chr == '*')
                    res = op2 * op1;
                else if (chr == '/')
                    res = op2 / op1;
                
                memo.push(res);
            }
        }
        return memo.top();
    }
};

File: Evaluate Reverse Polish Notation/Evaluate_Reverse_Polish_Notation.py
# 这里要注意python中的 // 运算符的问题所在

# Runtime: 36 ms, faster than 100.00% of Python3 online submissions for Evaluate Reverse Polish Notation.
# Memory Usage: 12.9 MB, less than 24.06% of Python3 online submissions for Evaluate Reverse Polish Notation.

class Solution:
    def evalRPN(self, tokens: 'List[str]') -> 'int':
        memo = []

        for token in tokens:
            if '0' <= token <= '9' or len(token) > 1:
                memo.append(int(token))
            else:
                op1 = memo.pop()
                op2 = memo.pop()

                if token == '+':
                    memo.append(op2 + op1)
                elif token == '-':
                    memo.append(op2 - op1)
                elif token == '*':
                    memo.append(op2 * op1)
                else:
                    memo.append(int(op2 / op1))
                
        return memo.pop();
